File: packages/api/app/routers/registry.py
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from app.services import registry_service

router = APIRouter(prefix="/registry", tags=["registry"])


class RegistryEntryRequest(BaseModel):
    provider: str
    id: str
    name: str
    description: str
    unit: str
    frequency: str
    geography: str
    tags: list[str] = []
    exampleYaml: str
    updatedAt: int | None = None


@router.get("/search")
async def search_registry(
    q: str = Query(default=""),
    provider: str | None = Query(default=None),
    geography: str | None = Query(default=None),
    limit: int = Query(default=20, ge=1, le=100),
):
    return await registry_service.search_registry_entries(q, provider, geography, limit)


@router.get("/{provider}/{source_id}")
async def get_registry_entry(provider: str, source_id: str):
    entry = await registry_service.get_registry_entry(provider, source_id)
    if not entry:
        raise HTTPException(404, detail=f"Registry entry '{provider}/{source_id}' not found")
    return entry


@router.post("")
async def create_registry_entry(req: RegistryEntryRequest):
    return await registry_service.create_registry_entry(req.model_dump())

File: packages/web/convex/registry.ts
import { mutation, query } from "./_generated/server";
import { v } from "convex/values";

function matchesQuery(entry: {
  sourceId: string;
  name: string;
  description: string;
  geography: string;
  tags: string[];
}, queryText: string) {
  const needle = queryText.trim().toLowerCase();
  if (!needle) return true;
  const haystack = [
    entry.sourceId,
    entry.name,
    entry.description,
    entry.geography,
    ...entry.tags,
  ].join(" ").toLowerCase();
  return haystack.includes(needle);
}

export const list = query({
  args: { limit: v.optional(v.number()) },
  handler: async (ctx, { limit }) => {
    const items = await ctx.db.query("dataSourceRegistry").collect();
    const sorted = items.sort((a, b) => b.updatedAt - a.updatedAt);
    return sorted.slice(0, limit ?? 100);
  },
});

export const search = query({
  args: {
    query: v.optional(v.string()),
    provider: v.optional(v.string()),
    geography: v.optional(v.string()),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, { query: queryText, provider, geography, limit }) => {
    let items = provider
      ? await ctx.db.query("dataSourceRegistry").withIndex("by_provider", (q) => q.eq("provider", provider)).collect()
      : await ctx.db.query("dataSourceRegistry").collect();

    if (geography) {
      items = items.filter((item) => item.geography === geography);
    }
    if (queryText) {
      items = items.filter((item) => matchesQuery(item, queryText));
    }

    const sorted = items.sort((a, b) => {
      const aName = a.name.toLowerCase();
      const bName = b.name.toLowerCase();
      if (queryText) {
        const q = queryText.toLowerCase();
        const aStarts = Number(aName.startsWith(q));
        const bStarts = Number(bName.startsWith(q));
        if (aStarts !== bStarts) return bStarts - aStarts;
      }
      return aName.localeCompare(bName);
    });

    return sorted.slice(0, limit ?? 20);
  },
});

export const get = query({
  args: { provider: v.string(), sourceId: v.string() },
  handler: async (ctx, { provider, sourceId }) => {
    return await ctx.db
      .query("dataSourceRegistry")
      .withIndex("by_provider_source", (q) => q.eq("provider", provider).eq("sourceId", sourceId))
      .first();
  },
});

export const create = mutation({
  args: {
    provider: v.string(),
    sourceId: v.string(),
    name: v.string(),
    description: v.string(),
    unit: v.string(),
    frequency: v.string(),
    geography: v.string(),
    tags: v.array(v.string()),
    exampleYaml: v.string(),
    updatedAt: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const now = Date.now();
    const existing = await ctx.db
      .query("dataSourceRegistry")
      .withIndex("by_provider_source", (q) => q.eq("provider", args.provider).eq("sourceId", args.sourceId))
      .first();

    const payload = {
      ...args,
      updatedAt: args.updatedAt ?? now,
      createdAt: existing?.createdAt ?? now,
    };

    if (existing) {
      await ctx.db.patch(existing._id, payload);
      return existing._id;
    }

    return await ctx.db.insert("dataSourceRegistry", payload);
  },
});
